core/src/main/java/io/crate/types/CollectionType.java

```diff
@@ -21,11 +21,98 @@
 
 package io.crate.types;
 
-public interface CollectionType {
+import io.crate.Streamer;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
 
-    DataType<?> innerType();
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
 
-    static DataType unnest(DataType dataType) {
+/**
+ * A type which contains a collection of elements of another type.
+ */
+public abstract class CollectionType extends DataType implements Streamer {
+
+    protected DataType<?> innerType;
+
+    /**
+     * Construct a new Collection type
+     * @param innerType May be null if the type is read later through the Streamer interface
```

Comment:
Ok, will keep it for now. IMHO the empty constructor is not good practice since it prevents you from using `final` fields and is very error-prone if you forget to initialize. Will put this on my TODO list.